### PR TITLE
MINOR: remove `checksumOrNull` and `isValid` from Record

### PR DESCRIPTION
**Pull request by chia7712**
Reference: https://github.com/apache/kafka/pull/10470#discussion_r608347559

This PR includes following task

1. rewrite the checksum of `DumpLogSegments`
2. remove `checksumOrNull` and `isValid` from `Record` (blocked by #10470) 

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/main/scala/kafka/tools/DumpLogSegments.scala [hunk 1]
```diff
@@ -270,7 +270,11 @@ object DumpLogSegments {
 
             var prefix = s"${RecordIndent} "
             if (!skipRecordMetadata) {
-              print(s"${prefix}offset: ${record.offset} isValid: ${record.isValid} crc: ${record.checksumOrNull}" +
+              val checksum = record match {
+                case r: AbstractLegacyRecordBatch => r.checksum().toString
+                case _ => "null"
+              }
```
Comment (ijuma) [inline core/src/main/scala/kafka/tools/DumpLogSegments.scala]: I think we should skip the checksum string altogether at record level for message format v2 and later.

File: core/src/main/scala/kafka/tools/DumpLogSegments.scala [hunk 2]
```diff
@@ -280,7 +284,7 @@ object DumpLogSegments {
               if (batch.magic >= RecordBatch.MAGIC_VALUE_V2) {
                 print(" sequence: " + record.sequence + " headerKeys: " + record.headers.map(_.key).mkString("[", ",", "]"))
               } else {
-                print(s" crc: ${record.checksumOrNull} isvalid: ${record.isValid}")
+                print(s" crc: $checksum isvalid: ${record.isValid}")
```
Comment (ijuma) [inline core/src/main/scala/kafka/tools/DumpLogSegments.scala]: Are we printing the checksum twice or am I missing something?

Comment (chia7712) [reply core/src/main/scala/kafka/tools/DumpLogSegments.scala]: yep, it printed the checksum twice if magic code is smaller than 2. Will remove the duplicate print.

File: core/src/main/scala/kafka/tools/DumpLogSegments.scala (hunk 1 above)
Comment (chia7712) [reply core/src/main/scala/kafka/tools/DumpLogSegments.scala]: sure

File: clients/src/main/java/org/apache/kafka/common/record/Record.java [hunk 3]
```diff
@@ -132,4 +121,16 @@
...
+    static void ensureValid(Record record) {
+        if (record instanceof AbstractLegacyRecordBatch) {
+            ((AbstractLegacyRecordBatch) record).ensureValid();
+        }
+    }
+
+    static boolean isValid(Record record) {
+        if (record instanceof AbstractLegacyRecordBatch) {
+            return ((AbstractLegacyRecordBatch) record).isValid();
+        } else return true;
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/record/Record.java]: Why is this change an improvement?

Comment (chia7712) [reply clients/src/main/java/org/apache/kafka/common/record/Record.java]: The method “isVaild” is meaningless for v2 and later. It seems to me it is not worth being a interface. If this change is overkill, alternative is that we can add default implementation to those stale methods. The v2 and later can skip implementation of stale methods.

Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/common/record/Record.java]: Polymorphism is a fine way to deal with that though. It's more confusing if we have a combination of polymorphism and pattern matching (well, the if/else version of it which doesn't include exhaustiveness checking).

I don't see an issue with having an implementation that always return true or is a no-op for v2. A default may be ok, but it introduces more coupling between the interface and implementations. It seems that `isValid` is only used by `DumpLogSegments` (outside of tests), so we can maybe remove that altogether. I would probably leave `ensureValid` as it was before this PR.

Comment (chia7712) [reply clients/src/main/java/org/apache/kafka/common/record/Record.java]: thanks for nice suggestions. will copy that.

File: core/src/main/scala/kafka/tools/DumpLogSegments.scala [hunk 4]
```diff
@@ -279,8 +279,10 @@ object DumpLogSegments {
 
               if (batch.magic >= RecordBatch.MAGIC_VALUE_V2) {
                 print(" sequence: " + record.sequence + " headerKeys: " + record.headers.map(_.key).mkString("[", ",", "]"))
-              } else {
-                print(s" crc: ${record.checksumOrNull} isvalid: ${record.isValid}")
+              }
+              record match {
+                case r: AbstractLegacyRecordBatch => print(s" crc: ${r.checksum}}")
```
Comment (ijuma) [inline core/src/main/scala/kafka/tools/DumpLogSegments.scala]: Shall we include the `isValid` part here too?

**Review by ijuma — COMMENTED** [review]
Looks good overall, just one comment below.

File: core/src/main/scala/kafka/tools/DumpLogSegments.scala (hunk 4 above)
Comment (chia7712) [reply core/src/main/scala/kafka/tools/DumpLogSegments.scala]: sure!

**Review by ijuma — APPROVED** [review]
LGTM, thanks!

**Pull request merged**